cmd/shortener/main.go

```diff
@@ -1,11 +1,57 @@
 package main
 
 import (
-	"github.com/UndeadDemidov/yandex-praktikum/internal/app"
+	"context"
 	"log"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/spf13/viper"
+
+	"github.com/UndeadDemidov/yandex-praktikum/internal/app/handlers"
+	"github.com/UndeadDemidov/yandex-praktikum/internal/app/server"
+	"github.com/UndeadDemidov/yandex-praktikum/internal/app/storages"
 )
 
+// ToDo Поленился вынести все в что-нибудь типа Execute()
+// Сделаю в первый же выходной перед 3-м спринтом
 func main() {
-	s := app.NewServer("http://localhost:8080/", ":8080")
-	log.Fatalln(s.ListenAndServe())
+	var s *http.Server
+	var r handlers.Repository
+	var err error
+
+	if r, err = storages.NewFileStorage(viper.GetString("file-storage-path")); err == nil {
+	} else {
+		r = storages.NewLinkStorage()
+		log.Print("In memory storage will be used")
+	}
+	s = server.NewServer(viper.GetString("base-url"), viper.GetString("server-address"), r)
+
+	// Пришлось сделать graceful shutdown, чтобы правильно закрывать файлы при остановке сервера
+	done := make(chan os.Signal, 1)
```

Comment:
это же отлично! кстати, здесь раз у нас версия выше 1.16 можно рассмотреть [вот такую конструкцию](https://henvic.dev/posts/signal-notify-context/)

Reply:
Переписал, не уверен что правильно схватил мысль автора, но получилось так:

```suggestion
	ctx, stop := signal.NotifyContext(context.Background(), os.Interrupt, syscall.SIGINT, syscall.SIGTERM)
	defer stop()

	go func() {
		if err := srv.ListenAndServe(); err != nil && err != http.ErrServerClosed {
			log.Fatalf("listen: %+v\n", err)
		}
	}()
	log.Print("Server started")

	if <-ctx.Done(); true {
		log.Print("Server stopped")
		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
		defer func() {
			err := repo.Close()
			if err != nil {
				log.Printf("Caught an error due closing file:%+v", err)
			}
			cancel()
		}()
		if err := srv.Shutdown(ctx); err != nil {
			log.Printf("Server Shutdown Failed:%+v", err)
		}
		stop()
		log.Print("Server exited properly")
	}
```

Reply:
верно) но только кажется, что false быть не может в условии ?

Reply:
Проглядел...

```
	<-ctx.Done()

	log.Print("Server stopped")
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer func() {
		err := repo.Close()
		if err != nil {
			log.Printf("Caught an error due closing file:%+v", err)
		}
		log.Println("Repository closed properly")
		cancel()
	}()
	if err := srv.Shutdown(ctx); err != nil {
		log.Printf("Server Shutdown Failed:%+v", err)
	}
	stop()
	log.Print("Server exited properly")
```

Копипаста враг мой :)
Все таки я еще не очень fluent с каналами, надо побольше практики, чтобы быстро подмечать такие нелогичности.